### PR TITLE
[CALCITE-1828] - Push the FILTER clause into Druid as a Filtered Aggregator

### PR DESCRIPTION
**Pull request by zhumayun**
### There were two ways to add support for factoring out filters from filter clauses (for data pruning and simplification):
1. Add a new `Filter` `RelNode` in between the current `RelNode`s in `DruidQuery`
2. Add a new `RexNode` member in `DruidQuery` that would have acted as a "alternative" filter.

I chose to go with option 1 because `DruidQuery` uses the `RelNode`s quite heavily in it's internal logic, and thus if I were to go with option 2 I would have had to add extra logic in those places to check if there was an "alternative" filter. This would have been messier (and more error prone) than the current approach.

### Instead of adding another Rule, `DruidAggregateProjectRule` was modified
There would have been little difference between a new rule and the old `DruidAggregateProjectRule`, so it made more sense to modify the existing rule. As you can see, most of the changes are just additional checks.

### An always false expression is pushed to Druid in some cases
In the `optimizeFilteredAggregations` method, sometimes the simplified expression will be `false`. Druid is unable to handle a filter that is simply "false", so the unsimplified filter to pushed into Druid, even though it's unsatisfiable. Normally when calcite encounters a always false filter, it doesn't use Druid, however in the optimization phase for filtered aggregations it's too late to go that route.

### CODE REVIEW

**Comment by zhumayun** [comment]
@julianhyde can you please review these changes? Thanks!

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java [hunk 1]
```diff
@@ -2156,6 +2156,308 @@ public RelNode apply(RelBuilder b) {
...
+  /**
+   * Test to ensure that multiple aggregates with filter clauses have their filters extracted to
+   * the outer filter field for data pruning
+   * */
+  @Test public void testFilterClausesFactoredForPruning1() {
+    String sql = "select "
+            + "sum(\"store_sales\") filter (where \"store_state\" = 'CA'), "
+            + "sum(\"store_sales\") filter (where \"store_state\" = 'WA') "
+            + "from \"foodmart\"";
+    String expectedQuery = "{'queryType':'timeseries','dataSource':'foodmart','descending':false,"
```
Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: Could we execute some of this queries, e.g., using ```returnsUnordered```, to verify that they are executed correctly?

Comment (zhumayun) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: Will do.

Comment (zhumayun) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: I've added code the execute the queries and check their results. Let me know if there's anything else you'd like me to add/remove.

Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: Thanks, LGTM.

**Comment by jcamachor** [comment]
@zhumayun , thanks! Could you rebase the patch? There are some changes conflicting with master.

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java [hunk 2]
```diff
@@ -2156,6 +2156,317 @@ public RelNode apply(RelBuilder b) {
...
+            + "sum(\"store_sales\") filter (where \"store_state\" = 'CA'), "
+            + "sum(\"store_sales\") filter (where \"store_state\" = 'WA') "
+            + "from \"foodmart\" where \"brand_name\" = 'Super'";
+    String expectedQuery = "{'queryType':'timeseries','dataSource':'foodmart','descending':false,"
+            + "'granularity':'all','filter':{'type':'and','fields':[{'type':'or','fields':[{'type':"
+            + "'selector','dimension':'store_state','value':'CA'},{'type':'selector','dimension':"
+            + "'store_state','value':'WA'}]},{'type':'selector','dimension':'brand_name','value':"
+            + "'Super'}]},'aggregations':[{'type':'filtered','filter':{'type':'selector',"
+            + "'dimension':'store_state','value':'CA'},'aggregator':{'type':'doubleSum','name':"
+            + "'EXPR$0','fieldName':'store_sales'}},{'type':'filtered','filter':{'type':'selector',"
```
Comment (b-slim) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: can we add at least one test where we have a nested filter ? like where store = ca or store = il

**Comment by zhumayun** [comment]
@b-slim I'ved added some test cases with more complicated filters. @jcamachor I don't see any merge conflicts on my end, i've merged the latest `master` branch into this branch as well. Do you still have issues on your end?

**Comment by jcamachor** [comment]
@zhumayun , you are right, it was my local master that was outdated. I will merge it now, thanks.

**Pull request closed without merging**